boards/opentitan/src/io.rs

```diff
@@ -0,0 +1,39 @@
+use core::fmt::Write;
+use core::panic::PanicInfo;
+use core::str;
+use kernel::debug;
+use kernel::hil::gpio;
+use kernel::hil::led;
+
+use crate::PROCESSES;
+
+struct Writer {}
+
+static mut WRITER: Writer = Writer {};
+
+impl Write for Writer {
+    fn write_str(&mut self, s: &str) -> ::core::fmt::Result {
+        debug!("{}", s);
```

Comment:
I think this might be a bit dangerous. The `debug!` mechanism is allowed to do some things to try to be pretty performant, e.g. on most platforms it layers on top of the console, which layers on top of DMA-based UART -- this is a design choice to lessen the (timing) impact of `debug!` as well as allowing it to multiplex with other users of the serial bus.

Once we've hit a panic, however, all bets are off. Generally, if a board supports printing during panic, it will implement a very simplistic byte-by-byte blocking write directly to the UART hardware here -- it's hard to be sure what else in the kernel is broken the case of a panic, so the idea is to minimize the amount of code that panic could end up touching. I'm not sure it's safe to use `debug!` inside the panic handler generally; I'm afraid once the blocking uart implementation is resolved, this will break (as the kernel scheduling loop is no longer running, so after the first print the debug buffer will never be returned to). For early board bringup, it may be easier to just blink LEDs

Reply:
I have fixed the print to follow what @alevy did here: https://github.com/tock/tock/pull/1478